azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go

```diff
@@ -0,0 +1,390 @@
+package datafactory
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/datafactory/mgmt/2018-06-01/datafactory"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmDataFactoryLinkedServiceOData() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmDataFactoryLinkedServiceODataCreateUpdate,
+		Read:   resourceArmDataFactoryLinkedServiceODataRead,
+		Update: resourceArmDataFactoryLinkedServiceWebCreateUpdate,
+		Delete: resourceArmDataFactoryLinkedServiceODataDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAzureRMDataFactoryLinkedServiceDatasetName,
+			},
+
+			"data_factory_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.DataFactoryName(),
+			},
+
+			// There's a bug in the Azure API where this is returned in lower-case
+			// BUG: https://github.com/Azure/azure-rest-api-specs/issues/5788
+			"resource_group_name": azure.SchemaResourceGroupNameDiffSuppress(),
+
+			"authentication_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(datafactory.ODataAuthenticationTypeAadServicePrincipal),
+					string(datafactory.ODataAuthenticationTypeAnonymous),
+					string(datafactory.ODataAuthenticationTypeBasic),
+					string(datafactory.ODataAuthenticationTypeWindows),
+				}, false),
+			},
+
+			"url": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"username": {
```

Comment:
I think this is going to cause headaches for users when trying to figure which values to pass in based on the authentication type they want to use. I think we can better help users there by doing something like the below for the schema
```
"anonymous_authentication": {
         Type:     schema.TypeList,
         MaxItems: 1,
	Optional: true,
        ExactlyOneOf: []string{"the", "other", "authentication", "methods"},
	Elem: &schema.Resource{
		Schema: map[string]*schema.Schema{
                 // fill out with the other details here
},
// Fill out the other authentication methods 
"service_principal_authentication": {
  ...
```

From that model, a user should be able to quickly see how best to setup authentication for this resource. What do you think?